packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts

```diff
@@ -315,14 +291,105 @@ function checkFunctionExpressionReturnType(
     node.type === AST_NODE_TYPES.ArrowFunctionExpression &&
     returnsConstAssertionDirectly(node)
   ) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Check that the function expression or declaration is valid.
+ */
+function isValidFunctionReturnType(
+  node:
+    | TSESTree.ArrowFunctionExpression
+    | TSESTree.FunctionDeclaration
+    | TSESTree.FunctionExpression,
+  options: Options,
+  isParentCheck = false,
+): boolean {
+  if (
+    !isParentCheck &&
+    options.allowHigherOrderFunctions &&
+    doesImmediatelyReturnFunctionExpression(node)
+  ) {
+    return true;
+  }
+
+  if (node.returnType || isConstructor(node.parent) || isSetter(node.parent)) {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Checks if a function declaration/expression has a return type.
+ */
+function checkFunctionReturnType(
+  node:
+    | TSESTree.ArrowFunctionExpression
+    | TSESTree.FunctionDeclaration
+    | TSESTree.FunctionExpression,
+  options: Options,
+  sourceCode: TSESLint.SourceCode,
+  report: (loc: TSESTree.SourceLocation) => void,
+): void {
+  if (isValidFunctionReturnType(node, options)) {
+    return;
+  }
+
+  report(getReporLoc(node, sourceCode));
+}
+
+/**
+ * Checks if a function declaration/expression has a return type.
+ */
+function checkFunctionExpressionReturnType(
+  node: TSESTree.ArrowFunctionExpression | TSESTree.FunctionExpression,
+  options: Options,
+  sourceCode: TSESLint.SourceCode,
+  report: (loc: TSESTree.SourceLocation) => void,
+): void {
+  if (isValidFunctionExpressionReturnType(node, options)) {
     return;
   }
 
   checkFunctionReturnType(node, options, sourceCode, report);
 }
 
+/**
+ * Check whether any ancestor of the provided node has a valid return type, with
+ * the given options.
+ */
+function ancestorHasReturnType(
+  ancestor: TSESTree.Node | undefined,
+  options: Options,
+): boolean {
+  let hasType = false;
+
+  while (ancestor) {
+    switch (ancestor.type) {
+      case AST_NODE_TYPES.ArrowFunctionExpression:
+      case AST_NODE_TYPES.FunctionExpression:
+        hasType =
+          isValidFunctionExpressionReturnType(ancestor, options) ||
+          isValidFunctionReturnType(ancestor, options, true);
+        break;
+      case AST_NODE_TYPES.FunctionDeclaration:
+        hasType = isValidFunctionReturnType(ancestor, options, true);
+        break;
+    }
+
+    ancestor = ancestor.parent;
+  }
```

Comment:
Almost correct, I can see two problems with this:

1. You should exit if you see anything that's not a `ReturnStatement`.

Otherwise the following code will match:
```ts
function a(): void {
  function b() {} // not returned
  const x = () => {};  // not returned
  return;
}
```

This is that it will return `false` for the following case:

```ts
function a() /* no return */ {
  return function b(): () => void {
    return function 
  }
}

```